cluster-autoscaler/cloudprovider/openstack/openstack_nodegroup.go

```diff
@@ -0,0 +1,288 @@
+	err = ng.WaitForClusterStatus(clusterStatusUpdateInProgress, waitForUpdateStatusTimeout)
+	if err != nil {
+		return fmt.Errorf("scale up error: %v", err)
+	}
+	err = ng.WaitForClusterStatus(clusterStatusUpdateComplete, waitForCompleteStatusTimout)
+	if err != nil {
+		return fmt.Errorf("scale up error: %v", err)
+	}
+	return nil
+}
+
+// DeleteNodes deletes a set of nodes chosen by the autoscaler.
+//
+// The process of deletion depends on the implementation of OpenstackManager,
+// but this function handles what should be common between all implementations:
+//   - simultaneous but separate calls from the autoscaler are batched together
+//   - does not allow scaling while the cluster is already in an UPDATE_IN_PROGRESS state
+//   - after scaling down, blocks until the cluster has reached UPDATE_COMPLETE
+func (ng *OpenstackNodeGroup) DeleteNodes(nodes []*apiv1.Node) error {
+
+	// Batch simultaneous deletes on individual nodes
+	ng.nodesToDeleteMutex.Lock()
+	// Check that these nodes would not make the batch delete more nodes than the minimum would allow
+	if *ng.targetSize-len(ng.nodesToDelete)-len(nodes) < ng.MinSize() {
+		ng.nodesToDeleteMutex.Unlock()
+		return fmt.Errorf("deleting nodes would take nodegroup below minimum size")
+	}
+	ng.nodesToDelete = append(ng.nodesToDelete, nodes...)
+	ng.nodesToDeleteMutex.Unlock()
+
+	ng.clusterUpdateMutex.Lock()
+	defer ng.clusterUpdateMutex.Unlock()
+
+	ng.nodesToDeleteMutex.Lock()
+	if len(ng.nodesToDelete) == 0 {
+		// Deletion was handled by another goroutine
+		ng.nodesToDeleteMutex.Unlock()
+		return nil
+	}
+	ng.nodesToDeleteMutex.Unlock()
+
+	// This goroutine has the clusterUpdateMutex, so will be the one
+	// to actually delete the nodes. While this goroutine waits, others
+	// will add their nodes to nodesToDelete and block at acquiring
+	// the clusterUpdateMutex lock. One they get it, the deletion will
+	// already be done and they will return above at the check
+	// for len(ng.nodesToDelete) == 0.
+	time.Sleep(deleteNodesBatchingDelay)
+
+	ng.nodesToDeleteMutex.Lock()
+	nodes = make([]*apiv1.Node, len(ng.nodesToDelete))
+	copy(nodes, ng.nodesToDelete)
+	ng.nodesToDelete = nil
+	ng.nodesToDeleteMutex.Unlock()
+
+	var nodeNames []string
+	for _, node := range nodes {
+		nodeNames = append(nodeNames, node.Name)
+	}
+	klog.V(1).Infof("Deleting nodes: %v", nodeNames)
+
+	updatePossible, currentStatus, err := ng.openstackManager.CanUpdate()
+	if err != nil {
+		return fmt.Errorf("could not check if cluster is ready to delete nodes: %v", err)
+	}
+	if !updatePossible {
+		return fmt.Errorf("can not delete nodes, cluster is in %s status", currentStatus)
+	}
+
+	minionCount, err := ng.openstackManager.NodeGroupSize(ng.id)
+	if err != nil {
+		return fmt.Errorf("could not get current node count")
+	}
+
+	if minionCount-len(nodes) < ng.MinSize() {
+		return fmt.Errorf("size decrease too large, desired:%d min:%d", minionCount-len(nodes), ng.MinSize())
+	}
+
+	var minionIPs []string
```

Comment:
I was not able to successfully target minions by IP when scaling down, I had to use the node index.  I attempted in Queens Heat with both Juno-era and Queens-era templates. When I inspected the cluster's minion ResourceGroup, I noticed that the "refs" attribute was null.  Would you be willing to share the output of the following API call?

curl -H 'X-Auth-Token:{OS_TOKEN}' {heat-api}/v1/{tenant_id}/stacks/{stack_name}/{stack_id}/resources/{resource_name}

Where resource_name is the kube_minions resourceGroup.  I am curious as to if the null refs attribute is the kicker for this.

Reply:
I wasn't able to find a list of what values are accepted by heat for minions_to_remove, and the only examples I saw were with the node index or IP, so I chose to use IP since it's directly available as a property of the node. I've now tried with the machine ID and that works as well for me. I won't be able to rework the tests for it and push the commit until Monday, but you can retest after that.

Reply:
we're not able to target the nodes by IP through this mechanism, even when passed through it just chooses the "newest" for downscaling. Only the index (maybe the machineID too? @scott-chaney) If you can show us the curl from that output we help us reduce our search space to have a better view of our problem.

Reply:
```
{
   "resource":{
      "resource_name":"kube_minions",
      "description":"",
      "links":[
         {
            "href":"https://openstack.cern.ch:8004/v1/d656bd69-82a2-4efe-bbee-abd0b0a83252/stacks/scaler-01-owjrxq5vkizf/5f3ad568-3260-4edb-91bb-c80dacaa275f/resources/kube_minions",
            "rel":"self"
         },
         {
            "href":"https://openstack.cern.ch:8004/v1/d656bd69-82a2-4efe-bbee-abd0b0a83252/stacks/scaler-01-owjrxq5vkizf/5f3ad568-3260-4edb-91bb-c80dacaa275f",
            "rel":"stack"
         },
         {
            "href":"https://openstack.cern.ch:8004/v1/d656bd69-82a2-4efe-bbee-abd0b0a83252/stacks/scaler-01-owjrxq5vkizf-kube_minions-g2tvz4jsooom/76467cd3-15db-43b8-a392-f8433e30d468",
            "rel":"nested"
         }
      ],
      "logical_resource_id":"kube_minions",
      "creation_time":"2019-02-15T13:25:01Z",
      "resource_status_reason":"state changed",
      "updated_time":"2019-02-15T13:50:40Z",
      "required_by":[

      ],
      "resource_status":"UPDATE_COMPLETE",
      "physical_resource_id":"76467cd3-15db-43b8-a392-f8433e30d468",
      "attributes":{
         "attributes":null,
         "refs":null,
         "refs_map":null,
         "removed_rsrc_list":[

         ]
      },
      "resource_type":"OS::Heat::ResourceGroup"
   }
}
```
I see refs as null as well.

Reply:
thank you!

Reply:
@tghartland Thanks for sharing the output, similar to ours.  From the [doc for ResourceGroup - removal_policies](https://docs.openstack.org/heat/rocky/template_guide/openstack.html)

> List of resources to be removed when doing an update which requires removal of specific resources. The resource may be specified several ways: (1) The resource name, as in the nested stack, (2) The resource reference returned from get_resource in a template, as available via the ‘refs’ attribute. Note this is destructive on update when specified; even if the count is not being reduced, and once a resource name is removed, its name is never reused in subsequent updates.

The node index works using the first path described.  I would think IP should fall in the second category.

Does the 'refs' attribute exist on any of the resources belonging to the ResourceGroup in your configuration?

@ricolin would you have any insight on ResourceGroup removal_policy?

Reply:
I have pushed a change to use the machineID to delete nodes. Let me know if it works for you.

Reply:
@scott-chaney please disregard my previous comment. Machine IDs do *not* work (they are just ignored by heat). IPs should work because heat is able to resolve them back to the node indices.
```
$ openstack stack show scaler-03-u6xxwz25daec-kube_minions-4ylgp5orn4qo
.....
| outputs               |
|                       |   output_key: refs_map
|                       |   output_value:
|                       |     '0': 188.185.64.117
.....
```
The solution is, as you said, to delete by the node index. Getting this from heat (without just parsing it from the node name) will be a bit of a pain but I'll start on that today. @strigazi and I have identified the change needed in the heat templates ([here](https://github.com/openstack/magnum/blob/d2390517518700d580283d29d09e851c60b78ebc/magnum/drivers/k8s_fedora_atomic_v1/templates/kubeminion.yaml#L547)) to instead map between index and machine ID in the stack output, which we think is more useful than having the IP there. If we get that change merged then it will be possible to delete via the machine IDs as well, which would simplify things here.

Reply:
@scott-chaney As for why IPs were failing for you, could you `describe` or `get node -o yaml` a node and compare the output for status.addresses to what I have?
```
status:
  addresses:
  - address: 188.185.64.117
    type: InternalIP
  - address: scaler-03-u6xxwz25daec-minion-0
    type: Hostname
```
And also check that the output corresponds to what you see in the outputs of the kube_minions stack, as I showed above.
```
$ openstack stack list --nested | grep scaler-03 | grep minions
| 48c8ebdf-5716-4dd5-acec-bec60838be0f | scaler-03-u6xxwz25daec-kube_minions-4ylgp5orn4qo-0-yv4zmqtad5go              | CREATE_COMPLETE | 2019-02-19T13:30:53Z | None                 | 99d3d801-9d6d-4f22-aa60-3bcab4e372bd |
| 99d3d801-9d6d-4f22-aa60-3bcab4e372bd | scaler-03-u6xxwz25daec-kube_minions-4ylgp5orn4qo                             | CREATE_COMPLETE | 2019-02-19T13:30:52Z | None                 | 91c102be-9b1c-4c12-a96f-b28b4c257cc8 |

$ openstack stack show scaler-03-u6xxwz25daec-kube_minions-4ylgp5orn4qo
........
|                       |   output_value:                                                                                           |
|                       |     '0': 188.185.64.117                                                                                   |
........

```

Reply:
@tghartland thanks for helping to debug.

When viewing the nested stack output, I see a `[resource_name]instance_id mapping`.  Your initial change of using the instance_id for scaling does appear to work for us.

This doesn't quite align with my expectations given the documentation for removal_policies on ResourceGroups.  I would think the resource should have the refs and refs_map (rather than just the nested stack having them) I am going to dig into the code a bit.

I am also not sure how our nested stacks have different resource mappings (yours with `[resource_name]ip_addr` and ours with `[resource_name]instance_id`).  

The default for a simple ResourceGroup appears to be `[resource_name]instance_id`. Do you somehow override this in your template or Heat configuration?

Reply:
@scott-chaney 
The mapping to ips comes from here [0] but instead of having [1] we have [2]. We don't have the last patch that changes it to the stack id. We are going to do a new change to use the nova server uuid [3] which is the most reasonable. If you have a mapping with instance uuids, you must be using [3].

[0] https://github.com/openstack/magnum/blob/master/magnum/drivers/k8s_fedora_atomic_v1/templates/kubeminion.yaml#L546-L549
[1] {get_param: "OS::stack_id"}
[2] {get_attr: [kube_minion_eth0, fixed_ips, 0, ip_address]}
[3] {get_resource: kube-minion}

Reply:
@scott-chaney to prevent being stuck between deleting by machine ID and it working for you, and deleting by IP and it working for @strigazi and myself, I have pushed a change that resolves back to the minion index using the nested kube_minions stack mapping, which checks both the ID and the IP.

https://github.com/kubernetes/autoscaler/pull/1690/commits/b3d231367b531464733d3af93fd1a00f3eb8a171#diff-d35c147e215a3b4712ad5f5cc79e6b82R356